instrumentation/opentelemetry-annotations-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/otelannotations/async/Jdk8MethodStrategy.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.otelannotations.async;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.tracer.BaseTracer;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+enum Jdk8MethodStrategy implements MethodSpanStrategy {
+  INSTANCE;
+
+  @Override
+  public boolean supports(Class<?> returnType) {
+    return returnType == CompletionStage.class || returnType == CompletableFuture.class;
+  }
+
+  @Override
+  public Object end(BaseTracer tracer, Context context, Class<?> returnType, Object result) {
+    if (supports(returnType)) {
+      if (result instanceof CompletableFuture) {
+        CompletableFuture<?> future = (CompletableFuture<?>) result;
+        if (endSynchronously(future, tracer, context)) {
+          return future;
+        }
+        return endWhenComplete(future, tracer, context);
+      } else if (result instanceof CompletionStage) {
+        CompletionStage<?> stage = (CompletionStage<?>) result;
+        return endWhenComplete(stage, tracer, context);
+      }
+    }
+    tracer.end(context);
+    return result;
+  }
+
+  /**
+   * Checks to see if the {@link CompletableFuture} has already been completed and if so
+   * synchronously ends the span to avoid additional allocations and overhead registering for
+   * notification of completion.
+   */
+  private boolean endSynchronously(
+      CompletableFuture<?> future, BaseTracer tracer, Context context) {
+
+    if (!future.isDone()) {
+      return false;
+    }
+
+    if (future.isCompletedExceptionally()) {
+      // If the future completed exceptionally then join to catch the exception
+      // so that it can be recorded to the span
+      try {
+        future.join();
+      } catch (Exception exception) {
+        tracer.endExceptionally(context, exception);
+        return true;
+      }
+    }
+    tracer.end(context);
+    return true;
+  }
+
+  /**
+   * Registers for notification of the completion of the {@link CompletionStage} at which time the
+   * span will be ended.
+   */
+  private CompletionStage<?> endWhenComplete(
+      CompletionStage<?> stage, BaseTracer tracer, Context context) {
+    return stage.whenComplete(
```

Comment:
If the stage is already complete I think this is guaranteed to be synchronous - I guess we can remove endSynchronously?

Reply:
I double checked to be sure and yes, `whenComplete` should always be synchronous, at least given how it's implemented in `CompletableFuture<T>`.  Checking and completing synchronously was more about optimizing away the need for the callback and the extra allocations that requires.  It's not observable, but it is cheaper/faster.  But if that's not worth the complexity I can remove it.